opal/mca/btl/base/btl_base_am_rdma.c

```diff
@@ -178,7 +178,7 @@ struct mca_btl_base_rdma_hdr_t {
     uint64_t context;
 
     /* registration handles (if required) */
-    uint8_t handle_data[];
```

Comment:
NO ! There is a drastic difference between these 2. Pout 1 char if you want but don't make this a pointer, it will break all the PML:/BTL code.

Reply:
Ok.  I can't quite parse your suggestion, though: "Pout 1 char" -- what do you mean by that?

Reply:
@jsquyres The pointer is not the same as the flexible array member. What is the warning that this is supposed to fix?

Reply:
I agree with George; this change is wrong, and this code looks legit.  I'd really be curious what warning is being thrown here.  It's possible that we're doing something in allocation to irritate the compiler?

Reply:
The warning is in the PR description:

```
  CC       base/btl_base_am_rdma.lo
base/btl_base_am_rdma.c:200:29: warning: invalid use of structure with flexible array member [-Wpedantic]
  200 |     mca_btl_base_rdma_hdr_t hdr;
      |                             ^~~

Reply:
Here's what the OMPI code is doing, reduced to a trivial example:

```
$ cat ~/tmp/foo.c
struct inner {
    int i;
    char a[];
};

struct outer {
    int j;
    struct inner i;
};
$ gcc ~/tmp/foo.c -c -pedantic
/home/jsquyres/tmp/foo.c:8:18: warning: invalid use of structure with flexible array member [-Wpedantic]
    8 |     struct inner i;
      |                  ^
```

The OMPI code has `struct inner` in the middle of an outer struct.  It apparently doesn't matter if inner is in the middle of the struct or is the last member of the struct, however.

Reply:
I'm trying to understand where the memory for the flexible array member is actually allocated but all I have found so far is casts from void pointers...

The flexible array member itself is never used but the memory seems to be accessed here: https://github.com/open-mpi/ompi/blob/master/opal/mca/btl/base/btl_base_am_rdma.c#L523 Maybe we can just remove the field? The struct is local to the TU so we're free to change it to whatever.

Reply:
The memory allocation happens on the peer, aka. whoever starts the RMA operation. Check [here](https://github.com/open-mpi/ompi/blob/master/opal/mca/btl/base/btl_base_am_rdma.c#L484) for an example.

I rechecked the code, and both, the code and the compiler complaint, are correct. The code is doing the right memory allocation and pointer manipulation, but using the flexible array in it's own struct, in the RDMA header. The compiler is also correct, because we encapsulate a structure with a flexible member inside a larger structure, something that is specifically not allowed by the C99 standard. However, we never use the flexible array in this case,  we keep this struct as a reminder of what the operation was, and whatever was in the flexible member is not stored in the `local_handle_data` and `remote_handle_data` fields in the `struct mca_btl_base_rdma_operation_t`.

We could avoid the compiler complain by splitting `struct mca_btl_base_rdma_hdr_t` in 2, a part that can be safely incorporated into another struct (aka. without the `handle_data` member) and the struct we have today which we will only use for network transfers. I think this is 1) not making the code more readable, and 2) should not be done by someone who does not maintain the above-mentioned code.

Reply:
In addition to what @bosilca just said ^^...

@bosilca and I chatted about this in Slack.  At least according to https://stackoverflow.com/questions/24108813/c99-flexible-array-inside-union, someone cited the C11 standard (ISO/IEC 9899:2011):

> A structure or union shall not contain a member with incomplete or function type [...], except that the last member of a structure with more than one named member may have incomplete array type; such a structure (and any union containing, possibly recursively, a member that is such a structure) shall not be a member of a structure or an element of an array.

The last phrase seems to state that having a nested struct where the inner struct contains a flex array is invalid.  However, `gcc` is only complaining about this when you use `-pedantic` (I tested with gcc 8.2.x and 10.2.x).  So this must be one of those places where gcc makes it work, even though it's technically invalid:

```
$ gcc foo.c -c
$ gcc foo.c -c -Wall
$ gcc foo.c -c -pedantic
foo.c:8:18: warning: invalid use of structure with flexible array member [-Wpedantic]
    8 |     struct inner i;
      |                  ^
```

I removed this change from the PR.